File: Week2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>

// count letters, words, sentences in a string
void count_string(char *string, int *letters, int *words, int *sentences) {
  // in_word is used to track whether we are currently in a word or not
  int in_word = 0;
  // loop through each character in the string
  for (char *letter = string; *letter != '\0'; letter++) {
    // if the character is a letter, increment the letter count and set in_word
    // to true
    if (isalpha(*letter)) {
      // increment the letter count and set in_word to true
      (*letters)++;
      in_word = 1;
    } else if (*letter == '.' || *letter == '!' || *letter == '?') {
      // if the character is a sentence-ending punctuation mark, increment
      // sentence count and set in_word to true
      (*sentences)++;
      in_word = 1;
    } else if (isspace(*letter)) {
      // if the character is a space, check if we are currently in a word. If we
      // are, increment the word count and set in_word to false
      if (!in_word) {
        in_word = 1;
      } else {
        in_word = 0;
        (*words)++;
      }
    }
  }
  // if we end the string while still in a word, we need to increment the word
  // count one last time
  if (in_word) {
    (*words)++;
  }
}

// calculate the Coleman-Liau index
int coleman_liau(int wc, float lc, float sc) {
  // calculate the average number of letters and sentences per 100 words
  float l_per_100 = (lc / wc) * 100;
  float s_per_100 = (sc / wc) * 100;
  // calculate the Coleman-Liau index using the formula
  float index = 0.0588 * l_per_100 - 0.296 * s_per_100 - 15.8;

  // round the index to the nearest whole number and return it
  return roundf(index);
}

// main function
int main(void) {
  // prompt the user for a string and initialize letter, word, and sentence
  // counts
  char *prompt = get_string("Text: ");
  int letters = 0;
  int words = 0;
  int sentences = 0;

  // count the letters, words, and sentences in the prompt
  count_string(prompt, &letters, &words, &sentences);

  // calculate the Coleman-Liau index and print the appropriate grade level
  int index = coleman_liau(words, letters, sentences);

  // print the appropriate grade level based on the index
  if (index < 1) {
    puts("Before Grade 1.");
  } else if (index > 16) {
    puts("Grade 16+.");
  } else {
    printf("Grade %d\n", index);
  }
}

File: Week6/readability.py
def count_string(prompt: str) -> tuple[int, int, int]:
    """Count the number of letters, words, and sentences in a given string."""
    letters = 0
    words = 0
    sentences = 0
    in_word = False
    # iterate through each character in the prompt
    for char in prompt:
        # check if the character is a letter
        if char.isalpha():
            # increment the letter count and set in_word to True
            letters += 1
            in_word = True
        # check if the character is a sentence-ending punctuation mark
        elif char in (".", "!", "?"):
            # increment the sentence count and set in_word to True
            sentences += 1
            in_word = True
        # check if the character is a whitespace character
        elif char.isspace():
            # if we are currently in a word, increment the word count and set in_word to False
            if in_word:
                in_word = False
                words += 1

    # if we ended the loop while still in a word, increment the word count
    if in_word:
        words += 1

    return letters, words, sentences


def coleman_liau(letters: int, words: int, sentences: int) -> int:
    """Calculate the Coleman-Liau index for a given number of letters, words, and sentences."""
    l_per_100 = (letters / words) * 100
    s_per_100 = (sentences / words) * 100
    return round(0.0588 * l_per_100 - 0.296 * s_per_100 - 15.8)


def main() -> None:
    prompt = input("Text: ")

    letters, words, sentences = count_string(prompt)

    cl_index: int = coleman_liau(letters, words, sentences)
    if cl_index < 1:
        print("Before Grade 1")
    elif cl_index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {cl_index}")


main()
